File: python/algorithms/sorting/heap_sort/heap_sort_inplace.py
# algorithms/sorting/heap_sort/heap_sort_inplace.py

from typing import List

def heapify(arr: List[int], n: int, i: int) -> None:
    
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr: List[int]) -> None:
    n = len(arr)
    # Step 1: Build max heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Step 2: Extract elements one by one
    for end in range(n - 1, 0, -1):
        # Move current root (max) to the end
        arr[0], arr[end] = arr[end], arr[0]
        # Restore heap property on reduced heap
        heapify(arr, end, 0)


if __name__ == "__main__":
    A = [7, 51, 4, 10, 15, 30, 50]
    heap_sort(A)
    print("Sorted ascending:", A)
    # Output: [4,7,10,15,30,50,51]

File: csharp/algorithms/sorting/heap_sort/HeapSortInPlace.cs
using System;

namespace Algorithms.Sorting;

public static class HeapSortInPlace
{
    // Function to maintain the max-heap property
    private static void Heapify(int[] arr, int n, int i)
    {
        int largest = i;           // root
        int left = 2 * i + 1;      // left child
        int right = 2 * i + 2;     // right child

        // If left child is larger than root
        if (left < n && arr[left] > arr[largest])
        {
            largest = left;
        }

        // If right child is larger than the largest so far
        if (right < n && arr[right] > arr[largest])
        {
            largest = right;
        }

        // If the largest is not root
        if (largest != i)
        {
            // Swap arr[i] with arr[largest]
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;

            // Recursively heapify the affected sub-tree
            Heapify(arr, n, largest);
        }
    }

    // Heap sort function
    public static void HeapSort(int[] arr)
    {
        int n = arr.Length;

        // Step 1: Build a max heap
        for (int i = n / 2 - 1; i >= 0; i--)
        {
            Heapify(arr, n, i);
        }

        // Step 2: Extract elements one by one
        for (int end = n - 1; end > 0; end--)
        {
            // Move current root (max element) to the end
            int temp = arr[0];
            arr[0] = arr[end];
            arr[end] = temp;

            // Call heapify on the reduced heap
            Heapify(arr, end, 0);
        }
    }

    // Demo
    /*public static void Main(string[] args)
    {
        int[] A = { 7, 51, 4, 10, 15, 30, 50 };
        HeapSort(A);
        Console.WriteLine("Sorted ascending: " + string.Join(", ", A));
        // Output: 4, 7, 10, 15, 30, 50, 51
    }*/
}
